hypervisor/persistence.go

```diff
@@ -145,46 +242,54 @@ func (pinfo *PersistInfo) vmContext(hub chan VmEvent, client chan *types.VmRespo
 		return nil, err
 	}
 
-	//ctx.vmSpec = pinfo.VmSpec
-	//ctx.userSpec = pinfo.UserSpec
-	//ctx.wg = wg
-
 	ctx.loadHwStatus(pinfo)
 
-	//for _, vol := range pinfo.VolumeList {
-	//	binfo := vol.blockInfo()
-	//	if len(vol.Containers) != len(vol.MontPoints) {
-	//		return nil, errors.New("persistent data corrupt, volume info mismatch")
-	//	}
-	//	if len(vol.MontPoints) == 1 && vol.MontPoints[0] == "/" {
-	//		img := &imageInfo{
-	//			info: binfo,
-	//			pos:  vol.Containers[0],
-	//		}
-	//		ctx.devices.imageMap[vol.Name] = img
-	//	} else {
-	//		v := &volume{
-	//			info:     binfo,
-	//			pos:      make(map[int]string),
-	//			readOnly: make(map[int][]bool),
-	//		}
-	//		for i := 0; i < len(vol.Containers); i++ {
-	//			idx := vol.Containers[i]
-	//			v.pos[idx] = vol.MontPoints[i]
-	//			v.readOnly[idx] = ctx.vmSpec.Containers[idx].RoLookup(vol.MontPoints[i])
-	//		}
-	//		ctx.devices.volumeMap[vol.Name] = v
-	//	}
-	//}
-	//
-	//for _, nic := range pinfo.NetworkList {
-	//	ctx.devices.networkMap[nic.Index] = &InterfaceCreated{
-	//		Index:      nic.Index,
-	//		PCIAddr:    nic.PciAddr,
-	//		DeviceName: nic.DeviceName,
-	//		IpAddr:     nic.IpAddr,
-	//	}
-	//}
+	ctx.networks.load(pinfo)
+
+	for _, vol := range pinfo.VolumeList {
+		binfo := vol.blockInfo()
+		ctx.volumes[binfo.Name] = &DiskContext{
+			DiskDescriptor: binfo,
+			sandbox:        ctx,
+			observers:      make(map[string]*sync.WaitGroup),
+			lock:           &sync.RWMutex{},
+			// FIXME: is there any trouble if we set it as ready when restoring from persistence
+			ready: true,
+		}
+	}
+
+	for _, pc := range pinfo.ContainerList {
```

Comment:
you might need to synthesize pinfo.VolumeList & pinfo.ContainerList from pinfo.VmSpec if it is being restored from the old saved data